File: 14次/t3 Midpoint method/test3.py

import matplotlib.pyplot as plt
import numpy as np
def pqt(p,q):
    temp=(p**2+q**2);
    return (-temp*q,temp*p)

def tixing(pi,qi,h):
    error=1e-14
    #为什么加大还崩了
    #不断迭代
    pi1=pi;qi1=qi;
    pi0=pi1+1;qi0=qi1+1;
    measure=1
    while(measure>error ):
    # for i in range(0,3):
        #存储
        pi0=pi1;qi0=qi1;
        #计算函数值
        pt,qt=pqt((pi+pi0)/2,(qi+qi0)/2)
        #迭代
        pi1=pi+h*pt;
        qi1=qi+h*qt;
        #一致收敛与我的区别
        measure=abs(pi1-pi0)+abs(qi1-qi0);
    return pi1,qi1

#中点公式
p=1;q=0

N=500000;
h=0.1;
x0=0;
x=[x0+h*i for i in range(0,N+1)];
p=np.zeros(N+1);
q=np.zeros(N+1)
p[0]=1;q[0]=0

for i in range(0,N):
    p[i+1],q[i+1]=tixing(p[i],q[i],h);
    # print(p[i+1])

plt.plot(p,q)

plt.xlabel('p',fontsize=20);
plt.ylabel('q',fontsize=20);
plt.title('Midpoint method',fontsize=20);
plt.tick_params(labelsize=20);

R=p**2+q**2;
maxR=max(R)
minR=min(R)
print("最大半径",maxR)
print("最小半径",minR)
print("半径误差",maxR-minR)

File: 14次/t3 Midpoint method/test3.m
clear

p0=1;
q0=0;

T=500000;
tau=0.1;



p=zeros(T,1);
q=zeros(T,1);

p(1)=p0;
q(1)=q0;



for t=2:T
    
    pm=p0;
    qm=q0;
    
    measure=1;
    while(measure>1e-14)
        p2=(pm+p0)/2;
        q2=(qm+q0)/2;
        
        pn=p0+tau*(-(p2^2+q2^2)*q2);
        qn=q0+tau*((p2^2+q2^2)*p2);
        
        measure=abs(pn-pm)+abs(qn-qm);
        pm=pn;
        qm=qn;
    end
    

    p(t)=pn;
    q(t)=qn;

    
    p0=pm;
    q0=qm;    
    

    
end


plot(p,q)

xlabel('p','Fontsize', 20);
ylabel('q','Fontsize', 20);
title('Midpoint method','Fontsize', 20);
set(gca, 'Fontsize', 20);

R=p.^2+q.^2;
maxR=max(R)
minR=min(R)

errorR=maxR-minR
